[WIP] Basic Type Validation

AtomCrafty opened the pull request:
This PR overhauls the way types are declared and handled internally, as the previous syntax and validation allowed many meaningless combinations like `short unsigned double`.

Valid types are:
```
void
bool

float
double

signed<SIZE>
unsigned<SIZE>

char    =   signed char     =   signed<8>
short   =   signed short    =   signed<16>
int     =   signed int      =   signed<32>
long    =   signed long     =   signed<64>

            unsigned char   =   unsigned<8>
            unsigned short  =   unsigned<16>
            unsigned int    =   unsigned<32>
            unsigned long   =   unsigned<64>
```

Closes #14

AtomCrafty committed:
Overhaul type specifier syntax

AtomCrafty committed:
Various grammar improvements

AtomCrafty committed:
Create new type classes

AtomCrafty committed:
Rewrite TypeProvider

AtomCrafty committed:
Adjust remaining classes

AtomCrafty committed:
Start fixing test cases

AtomCrafty committed:
Create integer type test case

jopperm commented:
(Assuming by the title that this is draft-PR; mark this as "ready to review" when I shall have a look!)

AtomCrafty commented:
Huh, weird. I was under the impression that adding WIP to the title automatically applies the "Draft" tag. Maybe that's a GitLab thing...

jopperm commented on com.minres.coredsl/src/com/minres/coredsl/typing/CompositeType.xtend:
```diff
@@ -0,0 +1,6 @@
+import org.eclipse.xtend.lib.annotations.Data;
+
+@Data class CompositeType extends DataType {
+}
```
Nit: Please add the EOF newlines (here and in the other new files).

jopperm reviewed (commented):
Did a quick pass over this, looks promising so far!

AtomCrafty committed:
Adjust integer literal terminal symbols

AtomCrafty committed:
Introduce parseAsX helpers

AtomCrafty committed:
Convert integer literals to TypedBigInteger

AtomCrafty committed:
Add integer literal tests

AtomCrafty committed:
Rename FloatingConstant to FloatConstant

AtomCrafty committed:
Implement type rules for arithmetic operators

AtomCrafty committed:
migrate from BigIntegerWithRadix to TypedBigInteger

AtomCrafty committed:
normalize line endings

AtomCrafty committed:
move issue codes to a separate class

AtomCrafty committed:
implement validation for assignments and declarations

AtomCrafty committed:
fix CoreDslIntegerTest

AtomCrafty committed:
simplify grammar

removed AbstractDeclarator
removed DirectAbstractDeclarator
removed UnaryOperator
removed Postfix

merged InitDeclarator and DirectDeclarator into Declarator

renamed Variable to Identifier

split PostfixExpression into FunctionCallExpression, ArrayAccessExpression and MemberAccessExpression
split PrimaryExpression into IdentifierReference, Constant and ParenthesisExpression

AtomCrafty committed:
rewrite logic for new grammar structure

AtomCrafty committed:
implement toString for types

AtomCrafty committed:
fix some test cases

AtomCrafty commented:
@jopperm @eyck Please review the grammar changes made in 38ba4ea308a7f9534dd9d976d0de0e08329d0b94.

I significantly simplified the declarator structure. Instead of `InitDeclarator`, `DirectDeclarator`, `AbstractDeclarator` and `DirectAbstractDeclarator` there now is only one `Declarator` symbol with an optional initializer clause.
I'm pretty sure we don't need any of the abstract declarators, as we never forward declare function signatures.

I've also rewritten `PrimaryExpression` and `PostfixExpression` such that they will generate dedicated classes for each syntax construct, instead of having multiple unrelated fields on one class.
For example: `42`, `x`, `"abc"` and `(1+2)` previously all generated a `PrimaryExpression` node with various fields set. Now, they will instead be parsed as `IntegerConstant`, `IdentifierReference`, `StringConstant` and `ParenthesisExpression` respectively.

The previous hierarchy of postfix operators was very unintuitive, as they were effectively parsed as if right associative, with the central expression as the root node. They are now instead parsed in a way that returns the last postfix operation as the root node, with each previous operator being nested within.

Example: `a.b->c` was previously parsed like this:
```
PostfixExpression
  left: PrimaryExpression
    ref: [a]
  postOp: Postfix
    op: "."
    declarator: [b]
    right: Postfix
      op: "->"
      declarator: [c]
```
But now it results in this parse tree:
```
MemberAccessExpression
  op: "->"
  declarator: [c]
  left: MemberAccessExpression
    op: "."
    declarator: [b]
    left: IdentifierReference
      identifier: [a]
```

eyck commented:
I don't like the PostfixExpression refactoring. In the current implementation the expression is represented left to right which is beneficial when generating code out of it. And I do not see that this is unintuitive.
What is the advantage of this refactoring?

AtomCrafty commented:
Expressions are typically evaluated from the inside outwards, i.e. the child nodes should be evaluated before the parent node is. 
All other operators follow this paradigm, only the postfix expression currently does the opposite.
It first evaluates the `left` subtree from the bottom up (as it should), and then proceeds to climb **down** the `postOp`/`right` subtree.

AtomCrafty commented:
Another key point: With the current structure, there are nodes that don't represent any sensible construct on their own.
With the new grammar, every node is a meaningful expression in its own right, and the code generation would be implemented using recursion, just as with basically all other structures.

![nesting](https://user-images.githubusercontent.com/4553050/145404395-5fdadb40-d5ec-4f0f-b166-5b232b7b1092.png)

jopperm commented:
I agree with @AtomCrafty regarding the postfix expression, the way the AST is currently built feels wrong. The value represented by `a.b->c` is the result of the `->` operator, which therefore should be the root node of the sub-AST. clang is doing exactly that.

I think an alternative closer to @eyck's mental model would be something like a "qualified identifier" or so, which could be parsed into a *list* of `.`/`->` separated components.

> With the new grammar, every node is a meaningful expression in its own right, and the code generation would be implemented using recursion, just as with basically all other structures.

That sounds like a desirable goal!
